### ISSUE TITLE
VS stack overflow crash on simple code with erroneous attributes

### ISSUE THREAD

**Issue by controlflow** [dotnet/roslyn#27228]
**Version Used**: 

15.8 preview

**Steps to Reproduce**:

Paste the following code into VS editor:
```c#
using System;
 
[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
[C(new C[0])] // err
[C(new string[0])]
[C(new string[1])] // err
[C(new string[1] { null })]
[C(new string[2] { null })]
public class C : Attribute {
  public C(C[] xs) { }
  public C(object[] xs) { }
 
  public static void Main() { }
}
```

**Expected Behavior**:

IDE is alive, compiler errors are reported.

**Actual Behavior**:

StackOverflowException somewhere inside Binder...

**Comment by jmarolf** [comment]
Was able to get a callstack
dump at \\mlangfs1\public\jmarolf\Dumps\so.dmp

```C#
Microsoft.CodeAnalysis.CSharp.Symbols.ArrayTypeSymbol.HasSameShapeAs(Microsoft.CodeAnalysis.CSharp.Symbols.ArrayTypeSymbol)
Microsoft.CodeAnalysis.CSharp.Symbols.ArrayTypeSymbol.Equals(Microsoft.CodeAnalysis.CSharp.Symbols.ArrayTypeSymbol, Microsoft.CodeAnalysis.TypeCompareKind)
Microsoft.CodeAnalysis.CSharp.Symbols.ArrayTypeSymbol.Equals(Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, Microsoft.CodeAnalysis.TypeCompareKind)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.ExpressionMatchExactly(Microsoft.CodeAnalysis.CSharp.BoundExpression, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterConversionFromExpression(Microsoft.CodeAnalysis.CSharp.BoundExpression, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, Microsoft.CodeAnalysis.CSharp.Conversion, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, Microsoft.CodeAnalysis.CSharp.Conversion, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef, Boolean ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterConversionFromExpression(Microsoft.CodeAnalysis.CSharp.BoundExpression, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, Microsoft.CodeAnalysis.CSharp.Conversion, Microsoft.CodeAnalysis.RefKind, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, Microsoft.CodeAnalysis.CSharp.Conversion, Microsoft.CodeAnalysis.RefKind, Boolean, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef, Boolean ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterFunctionMember[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.BoundExpression>, Boolean, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterFunctionMember[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.BoundExpression>, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.GetTheBestCandidateIndex[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>>, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.RemoveWorseMembers[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>>, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.ObjectCreationOverloadResolution(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol>, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol>, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.Binder.TryPerformConstructorOverloadResolution(Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, System.String, Microsoft.CodeAnalysis.Location, Boolean, Microsoft.CodeAnalysis.DiagnosticBag, Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol> ByRef, System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol> ByRef, Boolean)
Microsoft.CodeAnalysis.CSharp.Binder.BindAttributeConstructor(Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, Microsoft.CodeAnalysis.DiagnosticBag, Microsoft.CodeAnalysis.CSharp.LookupResultKind ByRef, Boolean, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.Binder.BindAttributeCore(Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.DiagnosticBag)
Microsoft.CodeAnalysis.CSharp.Binder.GetAttribute(Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.DiagnosticBag)
Microsoft.CodeAnalysis.CSharp.Binder.GetAttributes(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Binder>, System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax>, System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol>, Microsoft.CodeAnalysis.CSharp.Symbols.CSharpAttributeData[], Microsoft.CodeAnalysis.DiagnosticBag)
Microsoft.CodeAnalysis.CSharp.Symbol.LoadAndValidateAttributes(Roslyn.Utilities.OneOrMany`1<Microsoft.CodeAnalysis.SyntaxList`1<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax>>, Microsoft.CodeAnalysis.CustomAttributesBag`1<Microsoft.CodeAnalysis.CSharp.Symbols.CSharpAttributeData> ByRef, Microsoft.CodeAnalysis.CSharp.Symbols.AttributeLocation, Boolean, Microsoft.CodeAnalysis.CSharp.Binder, System.Func`2<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax,Boolean>)
Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetAttributesBag()
Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetAttributes()
// Call stack from here repeats below until stack overflow
Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.IsCustomTaskType(Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, System.Object ByRef)
Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.NormalizeTaskTypesInNamedType(Microsoft.CodeAnalysis.CSharp.CSharpCompilation, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol ByRef)
Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.NormalizeTaskTypesInType(Microsoft.CodeAnalysis.CSharp.CSharpCompilation, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol ByRef)
Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.NormalizeTaskTypesInArray(Microsoft.CodeAnalysis.CSharp.CSharpCompilation, Microsoft.CodeAnalysis.CSharp.Symbols.ArrayTypeSymbol ByRef)
Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.NormalizeTaskTypesInType(Microsoft.CodeAnalysis.CSharp.CSharpCompilation, Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol ByRef)
Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbolExtensions.NormalizeTaskTypes(Microsoft.CodeAnalysis.CSharp.Symbols.TypeSymbol, Microsoft.CodeAnalysis.CSharp.CSharpCompilation)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterFunctionMember[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.BoundExpression>, Boolean, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterFunctionMember[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>, Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.BoundExpression>, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.GetTheBestCandidateIndex[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>>, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.RemoveWorseMembers[[System.__Canon, mscorlib]](Microsoft.CodeAnalysis.PooledObjects.ArrayBuilder`1<Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<System.__Canon>>, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.OverloadResolution.ObjectCreationOverloadResolution(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol>, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol>, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.Binder.TryPerformConstructorOverloadResolution(Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, System.String, Microsoft.CodeAnalysis.Location, Boolean, Microsoft.CodeAnalysis.DiagnosticBag, Microsoft.CodeAnalysis.CSharp.MemberResolutionResult`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol> ByRef, System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol> ByRef, Boolean)
Microsoft.CodeAnalysis.CSharp.Binder.BindAttributeConstructor(Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.CSharp.AnalyzedArguments, Microsoft.CodeAnalysis.DiagnosticBag, Microsoft.CodeAnalysis.CSharp.LookupResultKind ByRef, Boolean, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef)
Microsoft.CodeAnalysis.CSharp.Binder.BindAttributeCore(Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.DiagnosticBag)
Microsoft.CodeAnalysis.CSharp.Binder.GetAttribute(Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax, Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol, Microsoft.CodeAnalysis.DiagnosticBag)
Microsoft.CodeAnalysis.CSharp.Binder.GetAttributes(System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Binder>, System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax>, System.Collections.Immutable.ImmutableArray`1<Microsoft.CodeAnalysis.CSharp.Symbols.NamedTypeSymbol>, Microsoft.CodeAnalysis.CSharp.Symbols.CSharpAttributeData[], Microsoft.CodeAnalysis.DiagnosticBag)
Microsoft.CodeAnalysis.CSharp.Symbol.LoadAndValidateAttributes(Roslyn.Utilities.OneOrMany`1<Microsoft.CodeAnalysis.SyntaxList`1<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeListSyntax>>, Microsoft.CodeAnalysis.CustomAttributesBag`1<Microsoft.CodeAnalysis.CSharp.Symbols.CSharpAttributeData> ByRef, Microsoft.CodeAnalysis.CSharp.Symbols.AttributeLocation, Boolean, Microsoft.CodeAnalysis.CSharp.Binder, System.Func`2<Microsoft.CodeAnalysis.CSharp.Syntax.AttributeSyntax,Boolean>)
Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetAttributesBag()
Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetAttributes()
```

**Comment by jcouv** [comment]
@gafter Would you be able to take a look at this one?

**Comment by chsienki** [comment]
Closed by https://github.com/dotnet/roslyn/pull/33498
